### PR TITLE
Make thumbnails for uploading pictures

### LINKED ISSUE #2138: Feature: Make thumbnail of upload image into channelData
[Edit by corinagum]: this issue has a workaround. Please see @EricDahlvang's comment at the bottom of this thread: 
> @handsomeyao77 Please see the workaround described here: [microsoft/botbuilder-dotnet#2143 (comment)](https://github.com/microsoft/botbuilder-dotnet/issues/2143#issuecomment-508830039)

--------
## Original comment:
Hi team,
I get an frequency error, the pictures can't be rendered in chat window, it's broken.

image

I view this picture src attribute in html:  
 src="https://bcdirectlineauseast.blob.core.windows.net/at27055/LSfhGXesbMy-original"

I think this is invalidate, because the right src in html is:
 src="https://directline.botframework.com/attachments/F7RM2rUmRwUI1uS3WxU8LF-0/0000113/0/win10-gbi-19.jpg?t=ew0KICAiYWxnIjogIlJTMjU2IiwNCiAgImtpZCI6ICJBT08tZXhGd2puR3lDTEJhOTgwVkxOME1tUTgiLA0KICAieDV0IjogIkFPTy1leEZ3am5HeUNMQmE5ODBWTE4wTW1ROCIsDQogICJ0eXAiOiAiSldUIg0KfQ.ew0KICAiYm90IjogImNoYXRib3Qtc2VhLWludCIsDQogICJzaXRlIjogImd0amo3anhFVWFjIiwNCiAgImNvbnYiOiAiRjdSTTJyVW1Sd1VJMXVTM1d4VThMRi0wLTAwMDAxMTMiLA0KICAibmJmIjogMTU2MjA0OTY3OSwNCiAgImV4cCI6IDE1NjIwNTE0NzksDQogICJpc3MiOiAiaHR0cHM6Ly9kaXJlY3RsaW5lLmJvdGZyYW1ld29yay5jb20vIiwNCiAgImF1ZCI6ICJodHRwczovL2RpcmVjdGxpbmUuYm90ZnJhbWV3b3JrLmNvbS8iDQp9.LKtl-Od1isNFau_40g1SMD4YAF6_qrI7sxC9qRXqWru5UJBzCSHfLWMKlzZbNk3yUMvagSbk2ThXkLxihuWeDIR3wMPVllsz7ubEiaIRqRuRhYtb3I8HpPiERmij1G4SSEar4ZOw1B7B5ugMK3bEf_0o1Ir4B6GNtOU0v6yMzbl52FZUzHjMVeHxDeLUdLF1m79YKWH5zaa74pX_yrXz0BQnahFlXRemAyL8IdKOuOHU7e9j4Zzag6Ca7Ump3xbBa9BCvAzrp_i28zlE7oM8qCzaShLX1iVNWLMcS0-YYGODedlILEOdnn3TCWdJGkPtFidNsQNI0q81fjBwsixOWg"

The webchat version is V3.

We have multiple chat dialog in the same time, if we switch between these dialog, the image message in chat window often broken.

image

## To Reproduce

1. Create more than 2 chat dialog
2. Chatting with any one(Dialog1), send text or image to him/her
3. Then switch to another one(Dialog2), send text or image  to him/her
4. Now switch back to 2# (Dialog1), the image which your sent is broken in chat window.

Please kindly help me on this issue, thanks very much!

### PR DESCRIPTION
**Pull request by compulim**
Fixes #2138.

## Changelog Entry

-  Make thumbnails for uploading GIF/JPEG/PNG and store it in `channelData.attachmentThumbnails`, by [@compulim](https://github.com/compulim), in PR [#2206](https://github.com/microsoft/BotFramework-WebChat/pull/2206), requires modern browsers with following features:
   - [Web Workers API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)
   - [`createImageBitmap`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/createImageBitmap)
   - [`MessageChannel`](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel)/[`MessagePort`](https://developer.mozilla.org/en-US/docs/Web/API/MessagePort)
   - [`OffscreenCanvas`](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas)
      - Specifically [`OffscreenCanvas.getContext('2d')`](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas/getContext)

#### Yesterday

When we attach a picture, Direct Line will echo back the activity with attachment of URL pointing to Direct Line temporary storage. Web Chat will replace the image with the URL echoed by Direct Line.

#### Today/tomorrow

When we attach a picture, Direct Line will **not** echo back the attachment URL, but a placeholder image. Since Web Chat replace the URL with a placeholder image, the UX is not great.

### Resolution

We are making thumbnail of every attaching picture, save it in `channelData.attachmentThumbnails`. As a bonus, since the thumbnail live inside the activity, it will survive <kbd>F5</kbd>.

By default, we are making thumbnail of 720x360 as JPEG of quality 0.6. It yields thumbnail of size around 20 KB with our testing images.

We detect whether the upload is a picture or not using filename extension (`.gif`/`.jpeg`/`.jpg`/`.png`). We are not using fingerprint of the file (e.g. `GIF87a`/`GIF89a`). **We do not support PDF files.**

Since thumbnail has a toll on the CPU, we are using few modern browser APIs to offload the work:

- [Web Workers API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API)
- [`createImageBitmap`](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/createImageBitmap)
- [`MessageChannel`](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel)/[`MessagePort`](https://developer.mozilla.org/en-US/docs/Web/API/MessagePort)
- [`OffscreenCanvas`](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas)
   - Specifically [`OffscreenCanvas.getContext('2d')`](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas/getContext)

The minimum requirement is Chrome 69 (for `OffscreenCanvas`). Although Firefox 46 support `OffscreenCanvas` thru an experimental flag, it do not support [`OffscreenCanvas.getContext('2d')`](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas/getContext). So this feature is disabled for Edge (UWP)/Firefox/Safari automatically thru capability check.

## Specific Changes

> Since thumbnail is a UI feature and requires specific user agent (browser) to work, we are adding it to `component` only. Devs using `core` will not have thumbnail creation.

- Modify `UploadButton.js`
   - `connectUploadButton.sendFiles` function updated to make thumbnail when picture is uploaded
- Added `downscaleImageToDataURL.js` and its singleton worker for offloading work
   - Send the `ArrayBuffer` to worker
   - Worker decode the picture, downscale it, encode it, and send it back as Base64 string
- Added `styleOptions` for toggling this feature, max width/height, image type, and image quality
- Added tests
   - WebDriver currently do not support uploading multiple files at the same time

---

-  [x] Testing Added

### CODE REVIEW

File: packages/component/src/Utils/downscaleImageToDataURL.worker.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
+// This file is the entrypoint of Web Worker and is minimally transpiled thru Babel.
```
Comment (corinagum) [inline packages/component/src/Utils/downscaleImageToDataURL.worker.js]: through

File: __tests__/setup/pageObjects/sendFile.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+function resolveLocalFile(filename) {
+  return join(__dirname, '../local', filename);
+}
+
+export default async function sendFile(driver, filename, { waitForSend = true } = {}) {
+  const uploadButton = await getUploadButton(driver);
+  const isUnderDocker = !!(await driver.getCapabilities()).get('webdriver.remote.sessionid');
+
+  // The send file function is asynchronous, it don't send immediate until thumbnails are generated.
```
Comment (corinagum) [inline __tests__/setup/pageObjects/sendFile.js]:
```suggestion
  // The send file function is asynchronous, it doesn't send immediately until thumbnails are generated.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -58,7 +58,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Added
 
--  Added bubble nub and style options, by [@compulim](https://github.com/compulim), in PR [#2137](https://github.com/Microsoft/BotFramework-WebChat/pull/2137)
+-  Added bubble nub and style options, by [@compulim](https://github.com/compulim), in PR [#2137](https://github.com/microsoft/BotFramework-WebChat/pull/2137)
+-  Make thumbnails for uploading GIF/JPEG/PNG and store it in `channelData.attachmentThumbnails`, by [@compulim](https://github.com/compulim), in PR [#2206](https://github.com/microsoft/BotFramework-WebChat/pull/2206), requires modern browsers with following features:
```
[the remark this answers is not in the document]
Comment (compulim) [reply CHANGELOG.md]: Couldn't believe I missed this.

File: packages/component/src/Utils/downscaleImageToDataURL.worker.js (hunk 1 above)
Comment (compulim) [reply packages/component/src/Utils/downscaleImageToDataURL.worker.js]: Fixed.

**Comment by compulim** [comment]
Talked to Chris and he actually suggest this (prefer Web Worker even it only works in Chrome for now).

I am unsure if he is aware that Firefox doesn't work yet.

**Comment by compulim** [comment]
Now support both Web Worker and non-Web Worker mode. Edge, Firefox, and IE11 will run in non-Web Worker mode.

Firefox has experimental (behind-the-flag) support for `OffscreenCanvas`, but it lack of encoding (`convertToBlob`). Instead of doing partial work in Web Worker, we will use non-Web Worker for Firefox.

**Comment by Sakshi0921** [comment]
@corinagum Do we need to update/change webchat or anything else from our side to see this fix in our bots?
We are still facing this problem in our projects.

**Comment by corinagum** [comment]
This bug wil be fixed in our next release in October. If you need these changes immediately, please see our documentation on pointing to our MyGet feed for [latest bits](https://github.com/microsoft/BotFramework-WebChat#how-to-test-with-web-chats-latest-bits).

**Pull request merged**